### ISSUE TITLE
CA2000 emitted when await using (false positive)

### ISSUE THREAD

**Issue by DrGriff** [dotnet/roslyn-analyzers#3042]
#### Analyzer package

[Microsoft.CodeAnalysis.FxCopAnalyzers](https://www.nuget.org/packages/Microsoft.CodeAnalysis.FxCopAnalyzers)

#### Package Version

v2.9.7 (Latest)

#### Diagnostic ID

Example: [CA2000](https://docs.microsoft.com/en-us/visualstudio/code-quality/ca2000?view=vs-2019)

#### Repro steps

1. .Net Core 3.1 (latest preview).  I happen to be using Blazor Server.
2. Create a Service class that accepts a DbContextOptions
3. Instantate a DbContext in a method.  Use the "await using..." C#8 syntax

```
public class MyService
{
	public DocTypesService(DbContextOptions<MyDbContext> options)
	{
		this.options = options;
	}

	private readonly DbContextOptions<MyDbContext> options;

	public async Task<Something> GetSomethingAsync(string id, CancellationToken token)
	{
		await using var context = new DocTypesContext(this.options);
			....
	}
}
```

#### Expected behavior

> await using var context = new DocTypesContext(this.options);

The context is created with a using statement, and will be disposed of when it goes out of scope and so no CA2000 error should **not** be emitted.

#### Actual behavior
> await using var context = new DocTypesContext(this.options);

A CA2000 error is emitted.  

However, were we to drop the `await` command `using var context = new DocTypesContext(this.options);` then the error is not emitted, but we really do need to put the `await` in.

**Comment by mscrivo** [comment]
Seeing the same behaviour after updating to VS 2019 16.4.  No way to fix, other than suppress the issue.

**Comment by Kharos** [comment]
Same here. For teams like ours who use code analysis and warnings as errors (we consider both "best practice"), this means that our existing solution is no longer building after upgrading Visual Studio. Not really a great way of introducing a new .net release 🐞

**Comment by DrGriff** [comment]
Similarly to @Kharos, we're also setting all warnings to Errors when compiling in Release mode.

It is a bit naff having to wrap the "offending" lines with pragma warning disable comments.

> #pragma warning disable CA2000 // Dispose objects before losing scope. Justification: False positive, see https://github.com/dotnet/roslyn-analyzers/issues/3042
>             await using var context = new MyContext(this.options);
> #pragma warning restore CA2000 // Dispose objects before losing scope

**Comment by mavasani** [comment]
https://github.com/dotnet/roslyn-analyzers/pull/3098 should fix this.

**Comment by mavasani** [comment]
Fixed with https://github.com/dotnet/roslyn-analyzers/pull/3098. I will post the beta package that you can try out to verify in a bit.

**Comment by mavasani** [comment]
https://dotnet.myget.org/feed/roslyn-analyzers/package/nuget/Microsoft.CodeAnalysis.FxCopAnalyzers/2.9.9-beta1.19610.1+906398e7 should have the fix.

**Comment by DenSmoke** [comment]
> https://dotnet.myget.org/feed/roslyn-analyzers/package/nuget/Microsoft.CodeAnalysis.FxCopAnalyzers/2.9.9-beta1.19610.1+906398e7 should have the fix.

When will the fix be available in release build?

**Comment by Yanal-Yves** [comment]
@DenSmoke, I have Just upgraded my project to "Microsoft.CodeAnalysis.FxCopAnalyzers" Version="3.0.0" and I don't have the issue anymore.

**Comment by GeeWee** [comment]
I still have it on 3.3.0 with the following code
```charp
await using var newMemoryStream = new MemoryStream();
```

**Comment by TaffarelJr** [comment]
I'm getting the false positives on `MemoryStream` as well as `GZipStream`.

I'm using `<EnableNETAnalyzers>true</EnableNETAnalyzers>` instead of FxCop, in a library targeting .NET Standard 2.1.

**Comment by daviddunson** [comment]
Just upgraded to VS 2022 and opened an existing project and got this warning on SQLiteConnection:

`                await using var command = this.Connection.CreateCommand();
`

It wants me to add ConfigureAwait() on a method that doesn't return a Task.

**Comment by daviddunson** [comment]
Please re-open this bug. I'm using VS 17.0.6 and experiencing the same problem with this code:

```
var lockStream = new FileStream(lockPath, FileMode.Open, FileAccess.Read, FileShare.ReadWrite);

await using (lockStream.ConfigureAwait(false))
{
}
```

"new FileStream" triggers CA2000.

**Comment by cytoph** [comment]
Yeah, the issue still exists. Please re-open this bug! It's super annoying that such a common feature is not handled correctly by the analyzers.
